desktop/src/main/java/bisq/desktop/main/offer/offerbook/OfferBook.java

```diff
@@ -83,44 +85,112 @@ public void onAdded(Offer offer) {
-                    OfferBookListItem offerBookListItem = new OfferBookListItem(offer);
-                    // We don't use the contains method as the equals method in Offer takes state and errorMessage into account.
-                    // If we have an offer with same ID we remove it and add the new offer as it might have a changed state.
-                    Optional<OfferBookListItem> candidateWithSameId = offerBookListItems.stream()
-                            .filter(item -> item.getOffer().getId().equals(offer.getId()))
-                            .findAny();
-                    if (candidateWithSameId.isPresent()) {
-                        log.warn("We had an old offer in the list with the same Offer ID. We remove the old one. " +
-                                "old offerBookListItem={}, new offerBookListItem={}", candidateWithSameId.get(), offerBookListItem);
-                        offerBookListItems.remove(candidateWithSameId.get());
+                    OfferBookListItem newOfferBookListItem = new OfferBookListItem(offer, hashOfPayload);
+                    removeDuplicateItem(newOfferBookListItem);
+                    offerBookListItems.add(newOfferBookListItem);  // Add replacement.
+                    if (log.isDebugEnabled()) {
+                        log.debug("onAdded: Added new offer {}\n"
+                                        + "\twith newItem.payloadHash: {}",
+                                offer.getId(),
+                                newOfferBookListItem.hashOfPayload == null ? "null" : newOfferBookListItem.hashOfPayload.getHex());
                     }
-
-                    offerBookListItems.add(offerBookListItem);
                 } else {
                     log.debug("We have the exact same offer already in our list and ignore the onAdded call. ID={}", offer.getId());
                 }
+                printOfferBookListItems("After onAdded");
             }
 
             @Override
-            public void onRemoved(Offer offer) {
-                removeOffer(offer, tradeManager);
+            public void onRemoved(Offer offer, P2PDataStorage.ByteArray hashOfPayload) {
+                printOfferBookListItems("Before onRemoved");
+                removeOffer(offer, hashOfPayload);
+                printOfferBookListItems("After onRemoved");
             }
         });
     }
 
-    public void removeOffer(Offer offer, TradeManager tradeManager) {
+    private void removeDuplicateItem(OfferBookListItem newOfferBookListItem) {
+        String offerId = newOfferBookListItem.getOffer().getId();
+        // We need to remove any view items with a matching offerId before
+        // a newOfferBookListItem is added to the view.
+        List<OfferBookListItem> duplicateItems = offerBookListItems.stream()
+                .filter(item -> item.getOffer().getId().equals(offerId))
+                .collect(Collectors.toList());
+        if (duplicateItems.size() > 0) {
```

Comment:
This check is not needed since `forEach` won't do anything if there are no items in the list.

Reply:
Removed list.size check in commit 4889da6bba6ad155436a2dede5182643f706ac6d.